src/tir/transforms/hoist_if_then_else.cc

```diff
@@ -0,0 +1,376 @@
+ * For example, given the following block:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * We first detect all IfThenElse stmt and find the corresponding loop invariant For stmt.
+ * Then we hoist IfThenElse stmt by one For stmt each step:
+ *
+ * Step 1:
+ * for (i = 0; i < 3; i++)
+ *     for (j = 0; j < 4; j++)
+ *         if (likely(i*2 < 4))
+ *             for (k = 0; k < 5; k++)
+ *                 A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * Step 2:
+ * for (i = 0; i < 3; i++)
+ *     if (likely(i*2 < 4))
+ *         for (j = 0; j < 4; j++)
+ *             for (k = 0; k < 5; k++)
+ *                 A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * In this pass, we only continue detecting possible hoisting chance when visiting For,
+ * IfThenElse or AttrStmt Node. For example, for the following block:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        A[i + j] = A[i + j] - 1
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * Only the For with k variable will be considered and the resulting stmt would be:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        A[i + j] = A[i + j] - 1
+ *        if (likely(i*2 < 4))
+ *            for (k = 0; k < 5; k++)
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *
+ * This pass doesn't do hoisting for consecutive IfThenElse stmt. The following
+ * block won't be optimized:
+ * for (i = 0; i < 3; i++)
+ *    for (j = 0; j < 4; j++)
+ *        for (k = 0; k < 5; k++)
+ *            if (likely(i*2 < 4))
+ *                A[3*i+2j+k] = B[7*i+3j+k]
+ *            if (likely(j > 2))
+ *                A[i+j+k] = B[i+j+k]
+ *
+ */
+
+// Select potential candidate IRs that can be hoisted.
+class HoistCandidateSelector final : public StmtExprVisitor {
+ public:
+  HoistCandidateSelector() { InitRecorder(); }
+
+  void VisitStmt_(const ForNode* op) final {
+    // Check if it is first for loop, then start the recorder
+    if (!RecordingComplete()) {
+      StartOrAddRecord(op);
+      StmtExprVisitor::VisitStmt_(op);
+      RemoveRecord(op);
+      return;
+    }
+
+    StmtExprVisitor::VisitStmt_(op);
+  }
+
+  void VisitStmt_(const SeqStmtNode* op) final {
+    // If SeqStmt is encountered in the middle of recording
+    //  then need to purge all, as it can not be hoisted
+    if (IsRecordingOn()) {
+      ResetRecorder();
+    }
+    StmtExprVisitor::VisitStmt_(op);
+  }
+
+  void VisitStmt_(const AttrStmtNode* op) final {
+    // Maintain list of all vars in AttrStmt
+    // To stop hoisting if any of the block variables are used.
+    //
+    // NOTE: If in future
+    // hoisting is required for any specific case,
+    // then add exception to only those case
+    // rather than allowing for all.
+    UpdateAttrVarList(op);
+    StmtExprVisitor::VisitStmt_(op);
+    RemoveAttrVarList(op);
```

Comment:
Thanks for bringing this up!
In fact i also had the same idea initially, and it was working perfectly fine.
But during CI failures i discovered, there are some dependency on the positioning of these If statements with global scope variables. So i have to put this logic to avoid hoisting for any such cases. 
But as i mentioned in the comment as well, if you have any specific case to enable hoisting, we can add it, provided it does not violate other Pass logic.

Please let me know your thought on this.

Reply:
Maybe the problem with that specific test is that some optimizations for threads require an exact positioning of the `if`s. Maybe we can put our `HoistIfThenElse` pass after that particular optimization pass.

Reply:
I found that the Passes working on these Attr nodes are distributed. Also the if nodes are likely optimized out in those cases.
But in reply to your concern, my current implementation does not stop hoisting for all Attr variables, it considers the required ones. It consider the if condition which has mixed case like `if(i + global_var ==  3)` .  So i think your actual concern is addressed.

But if you have any specific case, which current logic does not handle, please let me know. I can check on it. Thanks!

Reply:
Consider we split a loop `for (i = 0; i < n; i++)` into an inner loop and a outer loop, and bind the outer loop to `threadIdx.x`, there is common to be a condition like `if (threadIdx.x * tile_size + i.inner < n)`, where there are both `attr` variable `threadIdx.x` and loop variable `i.inner`. Can we handle this case?

Reply:
Thanks for you detail reply! Now i understand your point clearly.
Hypothetically if we have any such expression, yes it don't do hoisting. 

But i think these cases wont be there, because the loop splitting and fusion generates related blocks in a bound scope.
So in this scope we should not change the order of if and for nodes, where optimization technique solely depend on it.
I am not sure when the condition `If this condition is inside another loop, say Loop j, it should be hoisted.`
occurs. May be when we do out of order fusion i think, which is not a real-time use case.

However in case we really want to support this, may be we can add this Hoisting logic, only after all passes when all optimizations are done with a special config, which will scan only for such cases or we can add hoisting always after all the passes. I am not too sure about the benefit, maybe we can ask other's opinion too. Just to make sure all are in sync. :)

Reply:
I met those cases when I was working with sparse kernels. But I agree that we can merge this PR for now and left it for future improvements.

Reply:
I will check on sparse kernels once then. I will keep this point open. So that we can cover this scenario as well.

The solution is very simple, just move the position of pass to the end and disable the Attrs var list check feature. Unless there no further optimisation beyond lowering in the scope.